### ISSUE TITLE
How to find roots for Polynomial2D model?

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#3936]
@dthatte approached me with this problem:

She used `Polynomial2D` to do surface fitting of her image. We got to the point where we were able to generate a model image and access the coefficients (`c0_0`, etc.). She wants the peak location with sub-pixel precision (so just calling `np.where(model.max())` won't cut it). In her old code using Numpy's `poly1d`, she could use `numpy.roots` to find what she wants. Is there a similar method in `modeling` for this?

**Comment by embray** [comment]
I don't think there are any plans for the Polynomial models to provide all computations one might want to do with polynomials in general.  Though if someone had a proposal to add it that might be fine.  Adding a shortcut to `np.roots` would be easy enough though I don't know if it should stop there.
Note, as it is you can do:

```
np.roots(model.parameters)
```

**Comment by embray** [comment]
Actually, double-checking the docs that would have to be `np.roots(model.parameters[::-1])`.

**Comment by pllim** [comment]
Okay, I'll double check with her if this is what she wants. Thank you!

**Comment by dthatte** [comment]
Hi Erik and Pey-Lian,
I tried using np.roots as you suggested. I am not sure if it works in 2D.

Here is an example.

I want to fit a surface to the following array (image) and locate the peak.

The array 'im' looks like

array([[ 0.91528918,  0.92996254,  0.9429627 ,  0.94661716,  0.93425135],
       [ 0.92433812,  0.93942115,  0.95238806,  0.95545398,  0.94206077],
       [ 0.93340665,  0.94889253,  0.96237857,  0.96560326,  0.95169879],
       [ 0.93774628,  0.95339499,  0.96764186,  0.97153636,  0.9576178 ],
       [ 0.92527821,  0.94039141,  0.95488659,  0.95935463,  0.9459074 ]])

You can see the peak pixel has coordinates (4,4). I want to fit a surface and access the peak of the surface in fractional pixel units. I wanted to take derivative of the polynomial and see where it is zero.

Following the example on http://astropy.readthedocs.org/en/latest/modeling/ I did

p_init = models.Polynomial2D(degree=2)
fit_p=fitting.LinearLSQFitter()
p = fit_p(p_init, x, y, z)


To get the roots I did

np.roots(model.parameters[::-1])

It gives

array([ 10.10578261+2.63902106j,  10.10578261-2.63902106j,
         0.92552407+3.84233805j,   0.92552407-3.84233805j,  -3.02554249+0.j        ])

I don't see the root because the root should be between 4 and 5.


I think I need to use p.deriv and then np.roots following my one dimensional example below.

coefficients_x=np.polyfit(coords_x,vals_x,5)
polynomial=np.poly1d(coefficients_x)
roots_x=np.roots(np.polyder(coefficients_x))

However I don't know how to do this in 2D. Also I don't understand why p.deriv and p.get_num_coeff show the same output.

In [205]: p.deriv
Out[205]: <bound method Polynomial2D.deriv of <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001), c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0', value=-0.0047993498222241313), c0_1=Parameter('c0_1', value=0.016872893060568113), c0_2=Parameter('c0_2', value=-0.0033679519735650911), c1_1=Parameter('c1_1', value=0.00017691545063760051), param_dim=1)>>

In [206]: p.get_num_coeff
Out[206]: <bound method Polynomial2D.get_num_coeff of <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001), c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0', value=-0.0047993498222241313), c0_1=Parameter('c0_1', value=0.016872893060568113), c0_2=Parameter('c0_2', value=-0.0033679519735650911), c1_1=Parameter('c1_1', value=0.00017691545063760051), param_dim=1)>>

Could you please help?

Thanks
Deepashri
________________________________

**Comment by dthatte** [comment]
You should ask about this on the AstroPy mailing list where people might have 
more ideas:

http://mail.scipy.org/mailman/listinfo/astropy

On 07/17/2015 02:54 PM, Deepashri Thatte wrote:
> I tried using np.roots as you suggested. I am not sure if it works in 2D.

Rechecking the np.roots documentation, it appears it does not work for greater 
than 1D polynomials (sorry, I overlooked the fact that you were using 2D 
polynomials in your original message).

I don't know off hand a root finding implementation for 2D polynomials, but I'm 
sure it's out there (I'll look into it later).

Seems like getting the roots could actually be a useful addition to the models 
in Astropy, in general (where applicable).  There are several root finders in 
SciPy which could be supported in astropy.modeling.fitting:

http://docs.scipy.org/doc/scipy/reference/optimize.html#root-finding


> I think I need to use p.deriv and then np.roots following my one dimensional
> example below.
>
> coefficients_x=np.polyfit(coords_x,vals_x,5)
> polynomial=np.poly1d(coefficients_x)
> roots_x=np.roots(np.polyder(coefficients_x))
>
> However I don't know how to do this in 2D. Also I don't understand why p.deriv
> and p.get_num_coeff show the same output.
>
> In [205]: p.deriv
> Out[205]: <bound method Polynomial2D.deriv of
> <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001),
> c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0',
> value=-0.0047993498222241313), c0_1=Parameter('c0_1',
> value=0.016872893060568113), c0_2=Parameter('c0_2',
> value=-0.0033679519735650911), c1_1=Parameter('c1_1',
> value=0.00017691545063760051), param_dim=1)>>
>
> In [206]: p.get_num_coeff
> Out[206]: <bound method Polynomial2D.get_num_coeff of
> <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001),
> c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0',
> value=-0.0047993498222241313), c0_1=Parameter('c0_1',
> value=0.016872893060568113), c0_2=Parameter('c0_2',
> value=-0.0033679519735650911), c1_1=Parameter('c1_1',
> value=0.00017691545063760051), param_dim=1)>>

.deriv and .get_num_coeff are methods of the object "<bound method ..."  This is 
a function that you call that does something with the object (like ndarray.sum() 
or str.lower()).  I don't think they'll be of much help to you here though. 
Polynomial.deriv for example computes a jacobian matrix w.r.t. the parameters 
and is used for fitting routines--it's not the derivative of the polynomial 
itself.   This also indicates that the Astropy version you're using is over a 
year old, since the name of this method was changed to fit_deriv, to make this 
distinction clearer.  That said, upgrading Astropy won't offer anything for this 
specific problem.

**Comment by embray** [comment]
Rechecking the np.roots documentation, it appears it does not work for greater 
than 1D polynomials (sorry, I overlooked the fact that you were using 2D 
polynomials in your original message).

I don't know off hand a root finding implementation for 2D polynomials, but I'm 
sure it's out there (I'll look into it later).

Seems like getting the roots could actually be a useful addition to the models 
in Astropy, in general (where applicable).  There are several root finders in 
SciPy which could be supported in astropy.modeling.fitting:

http://docs.scipy.org/doc/scipy/reference/optimize.html#root-finding


> I think I need to use p.deriv and then np.roots following my one dimensional
> example below.
>
> coefficients_x=np.polyfit(coords_x,vals_x,5)
> polynomial=np.poly1d(coefficients_x)
> roots_x=np.roots(np.polyder(coefficients_x))
>
> However I don't know how to do this in 2D. Also I don't understand why p.deriv
> and p.get_num_coeff show the same output.
>
> In [205]: p.deriv
> Out[205]: <bound method Polynomial2D.deriv of
> <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001),
> c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0',
> value=-0.0047993498222241313), c0_1=Parameter('c0_1',
> value=0.016872893060568113), c0_2=Parameter('c0_2',
> value=-0.0033679519735650911), c1_1=Parameter('c1_1',
> value=0.00017691545063760051), param_dim=1)>>
>
> In [206]: p.get_num_coeff
> Out[206]: <bound method Polynomial2D.get_num_coeff of
> <Polynomial2D(c0_0=Parameter('c0_0', value=0.91210411317132001),
> c1_0=Parameter('c1_0', value=0.024392730808065199), c2_0=Parameter('c2_0',
> value=-0.0047993498222241313), c0_1=Parameter('c0_1',
> value=0.016872893060568113), c0_2=Parameter('c0_2',
> value=-0.0033679519735650911), c1_1=Parameter('c1_1',
> value=0.00017691545063760051), param_dim=1)>>

`.deriv` and `.get_num_coeff` are methods of the object "`<bound method ...`"  This is 
a function that you call that does something with the object (like `ndarray.sum()` 
or `str.lower(`)).  I don't think they'll be of much help to you here though. 
Polynomial.deriv for example computes a jacobian matrix w.r.t. the parameters 
and is used for fitting routines--it's not the derivative of the polynomial 
itself.   This also indicates that the Astropy version you're using is over a 
year old, since the name of this method was changed to `fit_deriv`, to make this 
distinction clearer.  That said, upgrading Astropy won't offer anything for this 
specific problem.

**Comment by pllim** [comment]
Dejavu :)

**Comment by embray** [comment]
Yeah, that was weird.  I didn't realize the e-mail I was responding to was a GitHub comment.

**Comment by pllim** [comment]
Just addressing this particular problem with finding the peak using 2D polynomial with `n=2`. For the given data, I obtained:
```python
>>> p
<Polynomial2D(2, c0_0=0.9121041155428579, c1_0=0.016872892828570796, c2_0=-0.0033679518571428442, c0_1=0.024392729285714433, c0_2=-0.00479934957142853, c1_1=0.0001769156000000332)>
```

Then, using the formula given in the documentation (http://astropy.readthedocs.org/en/latest/api/astropy.modeling.polynomial.Polynomial2D.html#astropy.modeling.polynomial.Polynomial2D), I derived the formulae for the peaks by hand by setting both derivatives w.r.t. `x` and `y` to 0:
![img_20150717_214010](https://cloud.githubusercontent.com/assets/2090236/8759829/e2f24262-2ccc-11e5-84f8-cd8f4caf74e2.jpg)

The using the derived formulae and coefficients from `p`, I calculated the sub-pixel location of the peak:
```python
>>> x_peak = (2 * p.c0_2 * p.c1_0 - p.c0_1 * p.c1_1) / (p.c1_1 * p.c1_1 - 4 * p.c0_2 * p.c2_0)
>>> y_peak = (-p.c1_0 - 2 * p.c2_0 * x_peak) / p.c1_1
>>> x_peak, y_peak
(2.5729093881268645, 2.5886754782132515)
```

To show where the calculated peak is:
```python
>>> import matplotlib.pyplot as plt
>>> plt.imshow(im, interpolation='none')
>>> plt.axvline(x_peak)
>>> plt.axhline(y_peak)
>>> plt.xlabel('X')
>>> plt.ylabel('Y')
>>> plt.title('im')
```
![untitled](https://cloud.githubusercontent.com/assets/2090236/8759864/30a8c2a0-2cce-11e5-8ab9-f4dfdc05022d.png)

This is close enough to the `max()` value that I am confident my basic calculus is still pretty solid:
```
>>> np.where(im == im.max())
(array([3]), array([3]))
```

**Comment by dthatte** [comment]
We also calculated this by hand. I wasn't sure how the coefficients were named and the terms they were associated with. You solution and the link that you provided makes this clear.
________________________________

**Comment by pllim** [comment]
@dthatte , glad to help. Until you find a proper function out there to do this for you, I think the best bet right now is to write your own function to use the `p.cN_N` coefficients to calculate your peak, similar to above. As long as you always use `Polynomial2D` with `n=2`, the equations should hold. And if you increase the `n`, you just have to re-calculate the formulae once (by setting the derivatives to 0 with extra terms) and adjust your function accordingly.

**Comment by pllim** [comment]
I think we can close this issue for now.

**Comment by dthatte** [comment]
I tried the peak finding method on a Gaussian to double-check that it finds the location of the peak when I already know it.
When I know the peak is at (20, 20) it finds it at (19.7009045106, 19.7009045106). If I move the peak away from the center then the error is more.

def f(x, y):
    return np.exp(-(x-20.0) ** 2 - (y-20.0) ** 2)

X, Y = np.mgrid[:40,:40]
surface = f(X, Y)

p_init = models.Polynomial2D(degree=2)
fit_p=fitting.LinearLSQFitter()
#fit_p = fitting.LevMarLSQFitter()
p = fit_p(p_init, X, Y, surface)


x_peak = (2 * p.c0_2 * p.c1_0 - p.c0_1 * p.c1_1) / (p.c1_1 * p.c1_1 - 4 * p.c0_2 * p.c2_0)
y_peak = (-p.c1_0 - 2 * p.c2_0 * x_peak) / p.c1_1


print x_peak, y_peak

19.7009045106 19.7009045106

________________________________

**Comment by pllim** [comment]
@dthatte , if you want to fit a Gaussian profile, maybe you are better off using `models.Gaussian2D` (http://astropy.readthedocs.org/en/latest/api/astropy.modeling.functional_models.Gaussian2D.html) instead of `models.Polynomial2D`. With `Gaussian2D`, you can pass in your initial guess of the center but you will have to switch to non-linear fitter:
```python
g_init = models.Gaussian2D(amplitude=1, x_mean=x0, y_mean=y0, x_stddev=1, y_stddev=1)
fit_g = fitting.LevMarLSQFitter()
g = fit_g(g_init, X, Y, surface)
```

**Comment by pllim** [comment]
Also, with `Gaussian2D`, you can get the fitted center directly from the model parameters, `g.x_mean` and `g.y_mean`.

**Comment by dthatte** [comment]
This works but you have to know that the surface you are trying to fit is a Gaussian. The reason for using gaussian function with models.Polynomial2D was to use it as a test function to see how
well models.Polynomial2D works on a surface with unknown profile.

So it seems it does create a fit for the Gaussian profile but not as good as Gaussian2D.
________________________________

**Comment by pllim** [comment]
Feel free to correct me if I am wrong, but I have not encountered any method that will do a good job of "blind" fitting. The more initial guess you can give the model, the better the fit will be. Passing in initial guess for Gaussian is usually something like this:
```python
i = np.where(data == data.max())
y0 = i[0][0]
x0 = i[1][0]
sigma_x = sigma_y = 1  # not very sure about this one
amplitude = data.max()

**Comment by pllim** [comment]
Same principle apply to passing in initial guess for `Polynomial2D` if you are not sure your data are Gaussian, although I do not know on top of my head what would be good guess values if you do not know your data profile.

**Comment by dthatte** [comment]
Giving initial guess is a good suggestion. I will also try it for Polynomial2D.
________________________________

**Comment by nden** [comment]
Note that the `LinearLSQFitter` provides an analytical solution, so there's no need to have initial guess for the parameters. Initial guess is only needed with the iterative algorithms of the non-linear fitters.

**Comment by dthatte** [comment]
Why doesn't it find the peak of the Gaussian correctly?  For a Gaussian with peak at (20,20) the calculated peak is (19.7009045106,19.7009045106). See the above discussion.

**Comment by embray** [comment]
@dthatte You're using a linear least squares fit to approximate a parabola to a non-linear gaussian profile, so I'm not surprised that there's some error.  Least squares doesn't care about where the peak is in the original data so long as the resulting parameters minimize the sum of least squares.  I'm especially not surprised given that the model is being evaluated on a discrete grid.  You're using an analytic solution for the x and y coordinates of the peaks, but passing in estimated coefficents and so not getting the exact solution you desire (though if evaluate the fitted model on a grid of pixels the peak is still at (20, 20)).

If you want a closer fit within fractions of a pixel you can use a more fine-grained grid like:

```python
X, Y = np.mgrid[:400:0.1,:400:0.1]
```

More fine-grained you make it the closer the fitted solution will be to the original data.  Adding additional weighting to points near the center might help too (though I tried this and found a bug in the weighted least squares code with 2D models, so I'll fix that).

**Comment by embray** [comment]
(I should add--this is just a guess but I think the reason you're getting a slight shift to the top left is that there is a greater difference between the gaussian and the parabola toward the tail, and least squares will tend to give more weight to greater outliers, hence shifting the fitted parabola a bit so that it fits points in that corner better (at the expense of fits in the other corners or the center).  That's why I think giving less weight to the tail / outer bounds might help).

**Comment by embray** [comment]
To drive the point home I think it's worth looking at a 1D version of the same problem.  This is more or less what's going on in 2D too:
```python
>>> def f(x): return np.exp(-(x-20.0)**2)
>>> x = np.arange(40)
>>> curve = f(x)
>>> pp_init = models.Polynomial1D(degree=2)
>>> fit_p = fitting.LinearLSQFitter()
>>> p = fit_p(pp_init, x, curve)
>>> x_peak = -pp.c1 / (2 * pp.c2)
>>> x_peak
19.700753170243608

```
Off by almost the same amount as in the 2D case.  Plotting it shows what's going on drastically:
```python
>>> plt.plot(x, curve, 'g-')
>>> plt.plot(x, pp(x), 'r--')
>>> plt.show()
```
![figure_1](https://cloud.githubusercontent.com/assets/676149/8832603/0bb5d3a4-3078-11e5-817c-ac0bba8a4c39.png)

In the effort to fit the parabola at the tail it has to be squashed down significantly (it also shifts to the left just because of the order of evaluation, but that's harder to see here since it's by such a small amount).  This is in part because the gaussian is so spiked in the center due to being evaluated on such a coarse grid.  Using a finer grid probably won't help with the "squashing" but it does minimize the shift.  To get rid of the squashing you have to give less (or even zero) weight to the tails.

**Comment by pllim** [comment]
@embray , in actual analysis, `X` and `Y` are pixel coordinates, and `surface` is the image data. In that case, fractional pixel sampling is not possible without some kind of interpolation on the original data first.

**Comment by pllim** [comment]
@embray , just want to point out that in your example above, `p` magically turned into `pp`.

**Comment by embray** [comment]
Fixed. Not sure what's going on there.

**Comment by dthatte** [comment]
That's correct. We tried np.ndimage.zoom but prefer not to use it since it is another interpolation.
________________________________

**Comment by pllim** [comment]
@dthatte , maybe `photutils` is a better option for you. See https://photutils.readthedocs.org/en/latest/photutils/morphology.html

**Comment by embray** [comment]
I know they're pixel coordinates, but the example was contrived in the first place.  Why are we trying to fit a polynomial, specifically?  I imagine there's a reason.  But it will fit better if you exclude the tails.

**Comment by embray** [comment]
@pllim We should have a PyLunch talk on photutils :D

**Comment by pllim** [comment]
@larrybradley , you want to give a PyLunch talk on `photutils`? :D

**Comment by dthatte** [comment]
The goal was to find peak of a cross-correlation function to find peak of PSF.



On Jul 23, 2015, at 10:36 AM, "Erik Bray" <notifications@github.com<mailto:notifications@github.com>> wrote:


I know they're pixel coordinates, but the example was contrived in the first place. Why are we trying to fit a polynomial, specifically? I imagine there's a reason. But it will fit better if you exclude the tails.

**Comment by larrybradley** [comment]
@dthatte If that's your use case, then you should check out the `image-registration` package for advanced methods:
   - Docs: http://image-registration.readthedocs.org/en/latest/
   - Code: https://github.com/keflavich/image_registration

From the docs:
"The most dangerous approach that should be avoided is that of fitting a gaussian to the peak of a cross-correlation image: this is the only other method that allows for measurements of the errors on the best-fit shift, but it is likely to be systematically wrong. The peak-fitting approach is unstable to fitting cross-correlated structure (which may be “beam-shaped”) instead of the cross-correlation shift peak (which may have effectively no shape because it is sub-pixel)."

**Comment by larrybradley** [comment]
BTW, if you are interested, I wrote a quadratic (degree=2) peak fitter using `Polynomial2D` for a recent JWST sprint:  https://github.com/spacetelescope/imgeom/blob/master/imgeom/fitting.py#L73

**Comment by embray** [comment]
@larrybradley That function looks like it will have the same issue @dthatte raised here, that the peak can be offset a bit by points far from the peak.  I think the best solution though is to exclude data from the fit that's outside the bounds of the paraboloid part (the work in #3909 will help with this too I think).

**Comment by larrybradley** [comment]
@embray Yes, indeed.  Ideally one should have a rough estimate of the peak and use only values out to a small radius from that estimate.  There's another function in that module that essentially does that (but it has a very specific application for rectified/unrectified images).

**Comment by embray** [comment]
Got it--in the future this is what bounding boxes should help with too (models where it's relevant will come with useful default bounding boxes, but it will also be settable).

**Comment by dthatte** [comment]
I didn't use Gaussian for the cross-correlation function. For that I used Polynomial2D. I chose a section with relatively smoother gradient between neighboring pixel values. The location of the peak shifts slightly (see the figures) depending upon the size (data points) of the section. I can choose a small radius but it will be at the expense of not having enough data points for the fit.

I will take a look at the links that you suggested.
________________________________
